cmd/lotus-sentinel/daemon.go

```diff
@@ -0,0 +1,252 @@
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:  "api",
+			Value: "1234",
+		},
+		&cli.BoolFlag{
+			Name:  "bootstrap",
+			Value: true,
+		},
+		&cli.StringFlag{
+			Name:  "config",
+			Usage: "specify path of config file to use",
+		},
+		&cli.StringFlag{
+			Name:  "import-snapshot",
+			Usage: "import chain state from a given chain export file or url",
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		ctx := context.Background()
+		{
+			dir, err := homedir.Expand(cctx.String("repo"))
+			if err != nil {
+				log.Warnw("could not expand repo location", "error", err)
+			} else {
+				log.Infof("lotus repo: %s", dir)
+			}
+		}
+
+		r, err := repo.NewFS(cctx.String("repo"))
+		if err != nil {
+			return xerrors.Errorf("opening fs repo: %w", err)
+		}
+
+		if cctx.String("config") != "" {
+			r.SetConfigPath(cctx.String("config"))
+		}
+
+		err = r.Init(repo.FullNode)
+		if err != nil && err != repo.ErrRepoExists {
+			return xerrors.Errorf("repo init error: %w", err)
+		}
+
+		if err := paramfetch.GetParams(lcli.ReqContext(cctx), build.ParametersJSON(), 0); err != nil {
+			return xerrors.Errorf("fetching proof parameters: %w", err)
+		}
+
+		snapshot := cctx.String("import-snapshot")
+		if snapshot != "" {
+			if err := ImportChain(ctx, r, snapshot, true); err != nil {
+				return err
+			}
+		}
+
+		genBytes := build.MaybeGenesis()
+		genesis := node.Options()
+		if len(genBytes) > 0 {
+			genesis = node.Override(new(modules.Genesis), modules.LoadGenesis(genBytes))
+		}
+
+		isBootstrapper := false
+		shutdownChan := make(chan struct{})
+		liteModeDeps := node.Options()
+		var api api2.SentinelNode
+		stop, err := node.New(ctx,
+			node.SentinelNodeAPI(&api),
+			// TODO this could happen in the above call to SentinelNodeAPI
+			node.Override(new(*events.Events), modules.NewEvents),
+
+			node.Override(new(dtypes.Bootstrapper), isBootstrapper),
+			node.Override(new(dtypes.ShutdownChan), shutdownChan),
+			node.Online(),
+			node.Repo(r),
+
+			genesis,
+			liteModeDeps,
+
+			node.ApplyIf(func(s *node.Settings) bool { return cctx.IsSet("api") },
+				node.Override(node.SetApiEndpointKey, func(lr repo.LockedRepo) error {
+					apima, err := multiaddr.NewMultiaddr("/ip4/127.0.0.1/tcp/" +
+						cctx.String("api"))
+					if err != nil {
+						return err
+					}
+					return lr.SetAPIEndpoint(apima)
+				})),
+			node.ApplyIf(func(s *node.Settings) bool { return !cctx.Bool("bootstrap") },
+				node.Unset(node.RunPeerMgrKey),
+				node.Unset(new(*peermgr.PeerMgr)),
+			),
+		)
+		if err != nil {
+			return xerrors.Errorf("initializing node: %w", err)
+		}
+
+		endpoint, err := r.APIEndpoint()
+		if err != nil {
+			return xerrors.Errorf("getting api endpoint: %w", err)
+		}
+
+		// TODO: properly parse api endpoint (or make it a URL)
+		return serveRPC(api, stop, endpoint, shutdownChan, int64(cctx.Int("api-max-req-size")))
+	},
+}
+
+func ImportChain(ctx context.Context, r repo.Repo, fname string, snapshot bool) (err error) {
```

Comment:
I'd try to deduplicate this with the 'normal' daemon

Reply:
Any preference for where this should live? In order for other packages to use it, we'll have to move it out of this file since it's in `package main` -- my first thought was the repo package.

Reply:
Would prefer not to have to make the user switch between binaries for this task. Let's find a new home for this function in a suitable lotus package.